tensor/shape.go

```diff
@@ -22,7 +22,9 @@ func (s Shape) TotalSize() int {
 	return ProdInts([]int(s))
 }
 
-func (s Shape) calcStrides() []int {
+// calcStrides has an additional optional argument, maks. It is used to mask out given dimensions
```

Comment:
I think there should be another function called `calcStridesWithMask`. `calcStrides` is called A LOT, might make sense to just split the functions